VEDA_OS01/utils.py

```diff
@@ -121,4 +123,76 @@ def parse(self, stream, media_type=None, parser_context=None):
         """
         Simply return a string representing the body of the request.
         """
-        return stream.read()
\ No newline at end of file
+        return stream.read()
+
+
+def get_or_create_course(course_id, course_hex=None, create_course_override=False):
+    """
+    Retrieves a course associated with course_hex, course_id or a creates new one.
+
+    Arguments:
+        course_id: course id identifying a course run
+        course_hex: studio_hex identifying course runs
+        create_course_override: Boolean. If a course ID and hex is supplied,
+            but no course is found, create the course anyway.
+
+    Details:
+     - If course_hex is there, try getting course with course_hex.
+     - Otherwise try making use of course_id to get the associated course
+       and if no course is associated with the course_id, try creating
+       a new course with course_name, institution, edx_classid and
+       local_storedir.
+     - If both a course_hex and course_id is supplied, and create_course_override
+       is false, don't do anything. If create_course_override is true, create
+       the course anyway. This is for stage, where we do not use an RDS instance
+       for the database. A course could exist in studio stage, but not exist in
+       veda stage (the stage machine was rebooted, and the db did not persist).
+       In this case, we want to continue as if the course existed, and not raise
+       any errors.
+    """
+
+    def _create_course(course_key, course_id, studio_hex=None):
+        course_name = '{org} {number}'.format(org=course_key.org, number=course_key.course)
+        course = Course.objects.create(
+            course_name=course_name,
+            institution=course_key.org,
+            edx_classid=course_key.course,
+            local_storedir=course_id,
+            yt_proc=False
+        )
+
+        if studio_hex:
+            setattr(course, 'studio_hex', studio_hex)
+
+        return course
+
+    if not course_hex:
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            return
+
+        course = Course.objects.filter(institution=course_key.org, edx_classid=course_key.course).first()
+        if course:
+            course_runs = course.course_runs
+            if course_id not in course_runs:
+                course_runs.append(course_id)
+                course.local_storedir = ','.join(course_runs)
+                course.save()
+        else:
+            course = _create_course(course_key, course_id)
+    else:
+        try:
+            course = Course.objects.get(studio_hex=course_hex)
```

Comment:
Maybe try to use the existing django `get_or_create` function to get rid of all these code

Reply:
👍